content/advent-2016/how-to-write-good-tests-for-solid-code.md

```diff
@@ -0,0 +1,219 @@
+
+```go
+// UseManager
+type UserManager struct {
+    notifier UserNotifier // Used to schedule emails
+    store    UserStore    // Used to persist users
+}
+
+// SignUp creates user account pending activation and sends the activation email.
+// It will return an error if creating the user failed  due to invalid details or problems with the UserStore.
+func (um *UserManager) SignUp(ctx context.Context, user User) (*User, error) {
+    // [...]
+}
+
+// [...]
+```
+
+As we can see, `UserManager` is dependent on the following two interfaces:
+
+```go
+// UserNotifier specifies the methods to schedule emails to be sent to certain user
+type UserNotifier interface {
+    RequestActivation(ctx context.Context, id string) error
+    RecoverPassword(ctx context.Context, id string) error
+}
+
+// UserStore specifies the methods required to manage user accounts
+type UserStore interface {
+    Find(ctx context.Context, id string) (*User, error)
+    Update(context.Context, User) error
+    Create(context.Context, User) (*User, error)
+}
+```
+
+Following the SOLID principles, we will create a function to initialise the `UserManager` injecting the dependencies:
+
+```go
+func NewUserManager(store UserStore, notifier UserNotifier) *UserManager {
+    return &UserManager{store: store, notifier: notifier}
+}
+```
+
+Once we have everything wired, the implementation of the UserManager methods could be something like this:
+
+```go
+func (um *UserManager) SignUp(ctx context.Context, user User) (*User, error) {
+    u, err := um.store.Create(ctx, user)
+    if err != nil {
+        return nil, err
+    }
+    um.notifier.RequestActivation(ctx, u.ID)
+    return u, nil
+}
+```
+
+## What should we test?
+
+When testing an object, it receiving and sending messages:
+
+- **Incoming messages** refer to calls to methods on the tested object.
+- **Outgoing messages** refers to method calls the tested object does ot it's dependencies.
+
+Following with our example, if we were to test the `SignUp` method, `SignUp` would be the incoming message while the calls to `um.store.Create` and `um.notifier.RequestActivation` would be outgoing messages.
+
+Furthermore, a message can be a Query or a Command:
+
+- Query messages return data without changing anything. e.g.: `UserStore.Find(id string) (*User, error)` would return a user without making any changes in the store.
+- Command messages modify data without returning any data. e.g.: `UserStore.Update(User) error` would make changes in the store without returning any new data.
+- Some commands might return some data, but we should be careful, cautious messages that return data and make modifications. We must ensure the changes are never hidden side-effects but required business logic. e.g.: `UserStore.Create(User) (*User, error)` will add a user to the store and must return the information about such user so we can get the ID of the user.
```

Comment:
"we should be careful, cautious messages that return data and make modifications" -> "we should be **careful and cautious of** messages that return data and make modifications"